scripts/demangle.py

```diff
@@ -1,104 +1,170 @@
-                with open(json_path+'/'+json_file_name, 'r') as f:
-                    json_content,flag ='',0
-                    for line in f:
-                        if flag:
-                            json_content+=line
-                        elif line[0]=='{':
-                            json_content+=line
-                            flag=1
-                    data = json.loads(json_content)
-            except:
-                print('ERROR: ',json_file_name, "Json file illegal")
+                data = json.load(f)
+            except json.JSONDecodeError:
+                warning(f" {json_file_name}  Json file illegal")
                 continue
-    for file_func_name in data:
-        # if split not matched, need to add a check
-        mangle_func_name = file_func_name.split(file_name)
-        if len(mangle_func_name)<2:
-            raise ValueError(f"ERROR: File '{file_name}' not found, may open wrong json file")
-        else:
-            mangle_func_name = mangle_func_name[1]
-            demangle_func_name = demangle(mangle_func_name)
-            # Add a check demangle is correct?
-            if mangle_func_name == demangle_func_name or demangle_func_name=='':
-                print('ERROR: '+mangle_func_name+' demangle incorrect or unable to demangle')
+        info(f"NUMBER OF FUNCTION: {len(data)}")
+        pre_file_path, code_content = "", ""
+        for cnt in range(len(data)):
+            file_func_name = ""
+            for file_func_name_ in data[cnt]:
+                file_func_name = file_func_name_
+
+            # if split not matched, need to add a check
+            splited_file_func_name = file_func_name.split(file_func_delim)
+            file_path, mangle_func_name = (
+                splited_file_func_name[0],
+                splited_file_func_name[1],
+            )
+            # if this is c file then the mangle does not exist
+            try:
+                demangle_func_name = cpp_demangle.demangle(mangle_func_name)
+            except ValueError:
+                demangle_func_name = mangle_func_name
+
+            if demangle_func_name == None:  # mangle_func_name == demangle_func_name
+                warning(f" {mangle_func_name} demangle incorrect or unable to demangle")
+                data[cnt][
+                    file_func_name
+                ] = f" {mangle_func_name}demangle incorrect or unable to demangle"
             else:
-                func_name = demangle_func_name.split('(')[0]
-                # print('looking for func ', func_name, ' in  file ', file_path)
-                func_content = extract_func_code(func_name,file_path)
-                print(func_content)
+                func_name = demangle_func_name.split("(")[0]
+                # Try to get code_content, the json file may provide the path or not
+                # only get code when pre_file_path!=file_path
+                if pre_file_path != file_path:
```

Comment:
What is `pre_file_path` and why it needs to be different from `file_path`?

Reply:
Because the function may come from different source code files.

Reply:
I see. I just realized that you only extract one file from the docker each time. A better way is to extract the whole project at once, as starting a docker could be expensive. Please add a TODO here, so we can fix that later.